core/src/main/java/com/novoda/noplayer/internal/exoplayer/ExoPlayerFacade.java

```diff
@@ -66,14 +70,67 @@ boolean isPlaying() {
         return exoPlayer != null && exoPlayer.getPlayWhenReady();
     }
 
+    boolean isPlayingAdvert() {
+        return isPlaying() && exoPlayer.isPlayingAd();
+    }
+
+    boolean isPlayingContent() {
+        return isPlaying() && !exoPlayer.isPlayingAd();
+    }
+
+    long mediaDurationInMillis() throws IllegalStateException {
+        assertVideoLoaded();
+        return exoPlayer.getDuration();
+    }
+
     long playheadPositionInMillis() throws IllegalStateException {
         assertVideoLoaded();
         return exoPlayer.getCurrentPosition();
     }
 
-    long mediaDurationInMillis() throws IllegalStateException {
+    long advertBreakDurationInMillis() {
         assertVideoLoaded();
-        return exoPlayer.getDuration();
+        Timeline currentTimeline = exoPlayer.getCurrentTimeline();
+        if (!isPlayingAdvert() || adsLoader.isAbsent() || !(currentTimeline instanceof SinglePeriodAdTimeline)) {
+            return 0;
+        }
+        SinglePeriodAdTimeline adTimeline = (SinglePeriodAdTimeline) currentTimeline;
+        Timeline.Period period = adTimeline.getPeriod(0, new Timeline.Period());
+
+        int currentAdGroupIndex = exoPlayer.getCurrentAdGroupIndex();
+        int advertCount = period.getAdCountInAdGroup(currentAdGroupIndex);
+
+        long advertBreakDurationInMicros = combinedAdvertDurationInGroup(period, advertCount);
+        return C.usToMs(advertBreakDurationInMicros);
+    }
+
+    long positionInAdvertBreakInMillis() {
+        assertVideoLoaded();
+        Timeline currentTimeline = exoPlayer.getCurrentTimeline();
+        if (!isPlayingAdvert() || adsLoader.isAbsent() || !(currentTimeline instanceof SinglePeriodAdTimeline)) {
+            return 0;
+        }
+        SinglePeriodAdTimeline adTimeline = (SinglePeriodAdTimeline) currentTimeline;
+        Timeline.Period period = adTimeline.getPeriod(0, new Timeline.Period());
+
+        int advertCount = exoPlayer.getCurrentAdIndexInAdGroup();
+
+        long playedAdvertBreakDurationInMicros = combinedAdvertDurationInGroup(period, advertCount);
+        return C.usToMs(playedAdvertBreakDurationInMicros) + playheadPositionInMillis();
+    }
+
+    private long combinedAdvertDurationInGroup(Timeline.Period period,
+                                               int numberOfAdvertsToInclude) {
+        int adGroupIndex = exoPlayer.getCurrentAdGroupIndex();
+        long advertBreakDurationInMicros = 0;
+        for (int i = 0; i < numberOfAdvertsToInclude; i++) {
+            long advertDurationInMicros = period.getAdDurationUs(adGroupIndex, i);
+            if (advertDurationInMicros == C.TIME_UNSET) {
+                advertDurationInMicros = adsLoader.get().advertDurationBy(adGroupIndex, i);
```

Comment:
can `adsLoader.get()` be extracted outside of the loop? so that its done only once

Reply:
Good point 💯

Reply:
https://github.com/novoda/no-player/pull/208/commits/65073331f1fc734fe2801ed4da75a899056deab3